include/zephyr/drivers/i3c.h

```diff
@@ -0,0 +1,1596 @@
+#define I3C_MSG_RW_MASK			BIT(0)
+/** @endcond  */
+
+/** Send STOP after this message. */
+#define I3C_MSG_STOP			BIT(1)
+
+/** RESTART I3C transaction for this message.
+ *
+ * @note Not all I3C drivers have or require explicit support for this
+ * feature. Some drivers require this be present on a read message
+ * that follows a write, or vice-versa.  Some drivers will merge
+ * adjacent fragments into a single transaction using this flag; some
+ * will not.
+ */
+#define I3C_MSG_RESTART			BIT(2)
+
+/** Transfer use HDR mode */
+#define I3C_MSG_HDR			BIT(3)
+
+/** I3C HDR Mode 0 */
+#define I3C_MSG_HDR_MODE0		BIT(0)
+
+/** I3C HDR Mode 1 */
+#define I3C_MSG_HDR_MODE1		BIT(1)
+
+/** I3C HDR Mode 2 */
+#define I3C_MSG_HDR_MODE2		BIT(2)
+
+/** I3C HDR Mode 3 */
+#define I3C_MSG_HDR_MODE3		BIT(3)
+
+/** I3C HDR Mode 4 */
+#define I3C_MSG_HDR_MODE4		BIT(4)
+
+/** I3C HDR Mode 5 */
+#define I3C_MSG_HDR_MODE5		BIT(5)
+
+/** I3C HDR Mode 6 */
+#define I3C_MSG_HDR_MODE6		BIT(6)
+
+/** I3C HDR Mode 7 */
+#define I3C_MSG_HDR_MODE7		BIT(7)
+
+/** I3C HDR-DDR (Double Data Rate) */
+#define I3C_MSG_HDR_DDR			I3C_MSG_HDR_MODE0
+
+/** I3C HDR-TSP (Ternary Symbol Pure-bus) */
+#define I3C_MSG_HDR_TSP			I3C_MSG_HDR_MODE1
+
+/** I3C HDR-TSL (Ternary Symbol Legacy-inclusive-bus) */
+#define I3C_MSG_HDR_TSL			I3C_MSG_HDR_MODE2
+
+/** I3C HDR-BT (Bulk Transport) */
+#define I3C_MSG_HDR_BT			I3C_MSG_HDR_MODE3
+
+/**
+ * @brief One I3C Message.
+ *
+ * This defines one I3C message to transact on the I3C bus.
+ *
+ * @note Some of the configurations supported by this API may not be
+ * supported by specific SoC I3C hardware implementations, in
+ * particular features related to bus transactions intended to read or
+ * write data from different buffers within a single transaction.
+ * Invocations of i3c_transfer() may not indicate an error when an
+ * unsupported configuration is encountered.  In some cases drivers
+ * will generate separate transactions for each message fragment, with
+ * or without presence of @ref I3C_MSG_RESTART in #flags.
+ */
+struct i3c_msg {
+	/** Data buffer in bytes */
+	uint8_t			*buf;
+
+	/** Length of buffer in bytes */
+	uint32_t		len;
+
+	/** Flags for this message */
+	uint8_t			flags;
+
+	/**
+	 * HDR mode (@c I3C_MSG_HDR_MODE*) for transfer
+	 * if @c I3C_MSG_HDR is set in @c flags.
+	 */
+	uint8_t			hdr_mode;
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * These are for internal use only, so skip these in
+ * public documentation.
+ */
+struct i3c_target_desc;
+struct i3c_i2c_device_desc;
+
+__subsystem struct i3c_driver_api {
+	int (*do_daa)(const struct device *dev);
+	int (*do_ccc)(const struct device *dev,
+		      struct i3c_ccc_payload *payload);
+
+	int (*i3c_xfers)(const struct device *dev,
+			 struct i3c_target_desc *target,
+			 struct i3c_msg *msgs,
+			 uint8_t num_msgs);
+	int (*i3c_device_register)(const struct device *dev,
+				   struct i3c_target_desc *desc);
+
+	int (*i3c_i2c_xfers)(const struct device *dev,
```

Comment:
Is it actually required by the driver to provide this transactional scatter/gather like functionality or can it provide a simpler subset with a helper that wraps the simpler subset of functionality here?

Reply:
Not necessarily... this is simply following the I2C API to make it a bit more familiar to developers.

Reply:
This API leaves it up to the driver to do polling/isr looping over each message transfer, while also ensuring the device is exclusively held from others.

The result is, at least in i2c, each driver basically repeats this same bit of looping logic and mutex/semaphore locking logic.

It means writing an async API each driver now has to reimplement the loop in an async way effectively replicating the logic again.

I'd at least strongly consider if it makes more sense to have the driver API provide a primitive for sending a single message with a callback rather than the entire message list. It would be relatively simple to build a i3c_transfer and i3c_transfer_async on top of that primitive. Leaving the exclusive access and looping logic in common code.

I get that might require some additional signaling to inform the driver to send stop and start signals rather having implicit or requiring look a heads in the list for RESTART. That doesn't seem like a significant issue from my perspective.

Reply:
Could you make a mock up of what you have in mind on the I2C side? I am not imaginative enough to picture what the whole solution would be.

Reply:
I can, it won't be until after next week though probably

Reply:
I can't quite picture what @teburd is saying either... but if the i3c_driver_api nests the i2c_driver_api it should be able to reuse all those `i2c_burst_write`, `i2c_write_read`, etc... This would also leave it open in the future for when/if an i2c async api is added to the `i2c_driver_api` it would just be a matter of implementing that api within the nested i2c_driver_api within the i3c controller
discussion here: https://github.com/zephyrproject-rtos/zephyr/pull/46087#discussion_r885023910